pydantic/generics.py

```diff
@@ -0,0 +1,76 @@
+from typing import TYPE_CHECKING, Any, Dict, Tuple, Type, TypeVar, Union, get_type_hints, no_type_check
+
+from pydantic import BaseModel, create_model
+from pydantic.class_validators import gather_validators
+
+
+class GenericModel:
+    __slots__ = ()
+    __parameters__ = ()
+
+    def __new__(cls, *args, **kwds):
+        if cls is GenericModel:
+            raise TypeError(f'Type {cls.__name__} cannot be instantiated; it can be used only as a base class')
+        else:
+            raise TypeError(f'Type {cls.__name__} cannot be instantiated without providing generic parameters')
+
+    if TYPE_CHECKING:  # pragma: no cover
+        def __init__(self, *args: Any, **kwargs: Any) -> None:
+            ...
+
+    @no_type_check
+    def __class_getitem__(cls, params: Union[Any, Tuple[Any, ...]]) -> Type[BaseModel]:
+        if not isinstance(params, tuple):
+            params = (params,)
+        if cls is GenericModel:
+            if not all(isinstance(x, TypeVar) for x in params):
+                raise TypeError(f'Each parameter to {cls.__name__} must be a TypeVar')
+            return type('ParameterizedGenericModel', (GenericModel,), {'__parameters__': params})
```

Comment:
I initially did this by declaring ParameterizedGenericModel as a class; this caused mypy to crash.